chain/indexer.go

```diff
@@ -408,6 +413,81 @@ func (t *TipSetIndexer) runProcessor(ctx context.Context, p TipSetProcessor, nam
 	}
 }
 
+// stateChangedActors is an optimized version of the lotus API method StateChangedActors. This method takes advantage of the efficient hamt/v3 diffing logic
+// and applies it to versions of state tress supporting it. These include Version 2 and 3 of the lotus state tree implementation.
+// stateChangedActors will fall back to the lotus API method when the optimized diffing cannot be applied.
+func (t *TipSetIndexer) stateChangedActors(ctx context.Context, old, new cid.Cid) (map[string]types.Actor, error) {
+	ctx, span := global.Tracer("").Start(ctx, "StateChangedActors")
+	if span.IsRecording() {
+		span.SetAttributes(label.String("old", old.String()), label.String("new", new.String()))
+	}
+	defer span.End()
+
+	var (
+		buf = bytes.NewReader(nil)
+		out = map[string]types.Actor{}
+	)
+
+	oldRoot, oldVersion, err := getStateTreeMapCIDAndVersion(ctx, t.node.Store(), old)
+	if err != nil {
+		return nil, err
+	}
+	newRoot, newVersion, err := getStateTreeMapCIDAndVersion(ctx, t.node.Store(), new)
+	if err != nil {
+		return nil, err
+	}
+
+	if newVersion == oldVersion && (newVersion == types.StateTreeVersion2 || newVersion == types.StateTreeVersion3) {
+		if span.IsRecording() {
+			span.SetAttribute("diff", "fast")
+		}
+		changes, err := hamt.Diff(ctx, t.node.Store(), t.node.Store(), oldRoot, newRoot, hamt.UseTreeBitWidth(5), hamt.UseHashFunction(func(input []byte) []byte {
+			res := sha256.Sum256(input)
+			return res[:]
+		}))
+		if err != nil {
+			log.Errorw("failed to diff state tree efficiently, falling back to slow method", "error", err)
+		} else {
```

Comment:
In hindsight, implementing a state tree shim and using here could make this less :spaghetti:.

Reply:
Probably a day or two worth of effort (including review). I would start by extracting a modified version of the [lotus StateTree](https://github.com/filecoin-project/lotus/blob/master/chain/state/statetree.go#L159) to visor and implement some code gen, similar to the actor-shims, to keep it up to date with the specs-actors actor-states version.